backend/src/ee/controllers/v3/machineIdentityController.ts

```diff
@@ -0,0 +1,726 @@
+    })
+}
+
+/**
+ * Return access token for machine identity with client id [clientId]
+ * and client secret [clientSecret]
+ * @param req 
+ * @param res 
+ */
+export const loginMI = async (req: Request, res: Response) => {
+    const {
+        body: {
+            clientId,
+            clientSecret
+        }
+    } = await validateRequest(reqValidator.LoginMachineIdentityV3, req);
+    
+    const machineIdentity = await MachineIdentity.findOne({
+        clientId,
+        isActive: true
+    });
+    
+    if (!machineIdentity) throw UnauthorizedRequestError();
+    
+    checkIPAgainstBlocklist({
+        ipAddress: req.realIP,
+        trustedIps: machineIdentity.clientSecretTrustedIps
+    });
+
+    const clientSecretData = await MachineIdentityClientSecretData.find({
+        machineIdentity: machineIdentity._id,
+        isActive: true
+    });
+    
+    let validatedClientSecretDatum: IMachineIdentityClientSecretData | undefined;
+    
+    for (const clientSecretDatum of clientSecretData) {
+        const isSecretValid = await bcrypt.compare(
+            clientSecret, 
+            clientSecretDatum.clientSecretHash
+        );
+
+        if (isSecretValid) {
+            validatedClientSecretDatum = clientSecretDatum;
+            break;
+        }
+    }
+    
+    if (!validatedClientSecretDatum) throw UnauthorizedRequestError();
+    
+    const {
+        expiresAt,
+        clientSecretUsageCount,
+        clientSecretUsageLimit
+    } = validatedClientSecretDatum;
+    
+    if (expiresAt && new Date(expiresAt) < new Date()) {
+        // client secret expired
+        await MachineIdentityClientSecretData.findByIdAndUpdate(
+            validatedClientSecretDatum._id,
+            {
+                isActive: false
+            },
+            {
+                new: true
+            }
+        );
+
+        throw UnauthorizedRequestError();
+    }
+    
+    if (clientSecretUsageLimit > 0 && clientSecretUsageCount === clientSecretUsageLimit) {
+        // number of times client secret can be used for 
+        // a login operation reached
+        await MachineIdentityClientSecretData.findByIdAndUpdate(
+            validatedClientSecretDatum._id,
+            {
+                isActive: false
+            },
+            {
+                new: true
+            }
+        );
+
+        throw UnauthorizedRequestError();
+    }
+
+    // increment usage count by 1
+    await MachineIdentityClientSecretData.findByIdAndUpdate(
+        validatedClientSecretDatum._id,
+        {
+            $inc: { clientSecretUsageCount: 1 }
+        },
+        {
+            new: true
+        }
+    );
+
+    // token version
+    const accessToken = createToken({
+        payload: {
+            machineId: machineIdentity._id.toString(), // consider changing to clientId and making it more extensible
+            clientSecretDataId: validatedClientSecretDatum._id.toString(),
```

Comment:
For `clientSecretDataId` I would do perhaps => `clientSecretId`; the `clientSecret` itself would be the actual value but the `clientSecretId` would be a reference to it.

`machineId` I would keep or put as `machineIdentityId` because it's similar to `user` and `userId`.

Reply:
I was thinking we might trip up and think that `clientSecretId` is the id of the related client secret but it's the reference to the document but up to you on naming